service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,372 @@
+
+        if (foundCertificates.isEmpty()) {
+            throw new GiftCertificateNotFoundException("Can't find gift certificates");
+        }
+
+        if (reqParams.containsKey(SORT_FIELDS_KEY)) {
+            String[] sortFields = reqParams.get(SORT_FIELDS_KEY);
+            String[] orders = reqParams.get(ORDER_KEY);
+
+            foundCertificates = sortingHelper.getSorted(sortFields, orders[0], foundCertificates);
+        }
+
+        return foundCertificates.stream()
+                .map(certificate -> modelMapper.map(certificate, GiftCertificateDto.class))
+                .peek(this::fillCertificateDtoWithTags)
+                .collect(Collectors.toList());
+    }
+
+
+    /**
+     * This method gets GiftCertificate entity from dao layer with given id and converts it to GiftCertificateDto.
+     * The method also calls {@link #fillCertificateDtoWithTags(GiftCertificateDto)} for filling gotten GiftCertificateDto
+     * with list of TagDto.
+     *
+     * @param id id of necessary entity.
+     * @return GiftCertificateDto with id and tags.
+     * @throws GiftCertificateNotFoundException if there is no entity with given id in database.
+     * @since 1.0
+     */
+    @Override
+    public GiftCertificateDto findById(long id) {
+        Optional<GiftCertificate> foundCertificateOpt = giftCertificateDao.getById(id);
+
+        GiftCertificate foundCertificate = foundCertificateOpt.orElseThrow(() ->
+                new GiftCertificateNotFoundException(String.format("Can't find a certificate with id: %d", id)));
+
+        GiftCertificateDto giftCertificateDto = modelMapper.map(foundCertificate, GiftCertificateDto.class);
+        fillCertificateDtoWithTags(giftCertificateDto);
+        return giftCertificateDto;
+    }
+
+
+    /**
+     * This method set date fields for GiftCertificate entity and saves it in database. Also the method link passed tags with the GiftCertificate
+     * entity and, if the tag doesn't exist in db, creates it.
+     *
+     * @param certificate GiftCertificate entity without date fields and id for saving.
+     * @return GiftCertificateDto with id and tags.
+     * @since 1.0
+     */
+    @Override
+    @Transactional
+    public GiftCertificateDto save(GiftCertificateDto certificate) {
+
+        String certificateName = certificate.getName();
+        Optional<GiftCertificate> foundCertOpt = giftCertificateDao.getByName(certificateName);//check existence of GiftCertificate
+        // in db by name
+
+        if (foundCertOpt.isPresent()) {
+            throw new GiftCertificateWithSuchNameAlreadyExists(String.format("Gift certificate with name: %s already exits.",
+                    certificate.getName()));
+        }
+
+        GiftCertificate giftCertificate = modelMapper.map(certificate, GiftCertificate.class);
+        String currentTime = DateHelper.getNowAsString();
+        giftCertificate.setCreateDate(currentTime);
```

Comment:
As I know there is a way to do this during inserting new record in DB through SQL
Does JDBCTemplate has such ability?

Reply:
I removed dateheler. Date is now handled by sql.
commits:  fa7af67 , 5e9a50c  (remove datehelper)